강원대 BE_오승규 4주차 과제 (2단계)

5win opened the pull request:
안녕하세요 제이님. 2단계 완료하여 PR드립니다!

말씀드렸던 "상품은 반드시 하나의 옵션을 가진다" 요구사항도 만족하도록 구현해보았습니다.
이 때문에 다소 코드가 복잡해진 감이 없지않아 있네요..

몇 가지 질문사항은 코드에 코멘트로 남겨두도록 하겠습니다!

5win committed:
feat: Add product option function
- Option 엔티티의 CRUD 기능
- Option과 Product 엔티티를 N:1 연관관계 매핑
- 관련 쿼리와 테스트 코드 추가.

5win committed:
feat: Add option input validation function and test code

5win committed:
test: Add entity test code

5win committed:
feat: Changed to add options when adding products
- 상품 추가할 때 옵션도 함께 추가가 가능하도록 변경.

5win committed:
fix: Add getter methods to argument resolving

5win committed:
feat: Must have an option when add a product
- 상품 추가 시에 옵션을 반드시 1개 이상 가지도록 제한.
- 옵션을 삭제할 경우에 대해선 아직 적용되지 않음.

5win committed:
feat: Can not delete option if only one option exists

5win committed:
feat: Add error message used to if category not be selected.

5win committed:
test: Modifying test code because the product must have an option

5win committed:
refact: Remove duplicate logic by extracting method in ProductService

5win committed:
refact: Remove duplicate logic by extracting method in MemberService

5win committed:
refact: Extract method from logic in productService

5win committed:
chore: Change parameter type

5win committed:
docs: Update README.md

5win commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -24,14 +24,14 @@ public ProductController(ProductService productService) {
     @PostMapping()
-    public ResponseEntity<String> addProduct(@RequestBody @Valid ProductDto product) {
-        productService.addProduct(product);
+    public ResponseEntity<String> addProduct(@RequestBody @Valid ProductRequest productRequest) {
```
상품을 등록할 때 옵션들을 추가해서 함께 전송하도록 했습니다.
이 과정에서 상품 등록 요청에는 옵션 객체의 정보가 포함되지만, 상품 리스트를 보여주는 응답의 경우에는 옵션 객체 데이터를 함께 보내지 않도록 했기 때문에 request와 response를 따로 분리했습니다.

각 상품의 옵션에 대한 정보는 상품 리스트에서 "옵션 목록" 을 눌렀을 때 따로 요청과 응답이 발생하도록 했기 때문에
ProductResponse에는 포함될 필요가 없다고 판단했습니다.

jaymarystella replied:
잘해보셨습니다. 보통 서브정보는 ui측면으로 분리를 해놓기도합니다. 사용자가 보는화면에 대한 사용성과, 서버부하 둘다 잡을 수 있는 방법으로 종종 쓰입니다.

5win commented on src/main/java/gift/controller/ProductViewController.java:
```diff
@@ -32,7 +32,7 @@ public ProductViewController(ProductService productService, CategoryService cate
      */
     @GetMapping()
     public String getAllProducts(Model model, @PageableDefault(size = 5) Pageable pageable) {
```
저는 상품 목록에서 각 상품에 대해 "옵션 목록"을 클릭하면 옵션 목록 페이지를 전달하도록 구성했는데요
만약 옵션 목록 페이지 따로 만들지 않고, 토글을 누르면 옵션이 보이도록 하는 식으로 만드려고 한다면 model에 옵션 리스트에 대한 정보도 포함시켜서 뷰를 전송해야 한다고 생각이 됩니다.

그런데 만약 목록 개수가 매우 많거나, 사진과 같은 용량이 큰 데이터가 포함되거나, 또는 데이터를 포함시키기까지의 쿼리 비용이 오래걸린다면 모든 데이터들이 사용될지도 모르는데 한 번에 포함시켜 보내는건 다소 비효율적일 것 같은데, 
실제로 최적화를 시켜서 사용한다고 하면 토글 버튼을 눌렀을 때 이벤트를 캐치해서 해당 토글에서 보여주어야할 데이터들을 다시 요청하는 식으로 설계는 하는 것이 좋을까요?

jaymarystella replied:
제가 프런트엔드 토글 컴포넌트에서 서버에 쏘는 방식을 잘 모르긴하지만, 브라우저에서 가지고있다가 서버통신없이 토글에서 그려주는걸 의미하신건가요? 

데이터의 양이 얼마안되면 상관없겠는데 필요할때 호출하는 방식이 더 나아보이긴합니다.
그리고 옵션이 매우많으면 당연히 다 내려주는건 비효율입니다. 페이지형태로 내려주거나 옵션필터 등을 적용해서 서버에서 필요한 정보만을 가지고오게하는방향으로 구현해야합니다.

5win replied:
> 제가 프런트엔드 토글 컴포넌트에서 서버에 쏘는 방식을 잘 모르긴하지만, 브라우저에서 가지고있다가 서버통신없이 토글에서 그려주는걸 의미하신건가요?


넵 말씀하신대로 의미한게 맞습니다.
페이지 형태로 보여줄 수도 있겠네요 감사합니다 :)

5win commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -33,47 +33,67 @@ public MemberService(MemberJpaDao memberJpaDao, WishlistJpaDao wishlistJpaDao,
+        memberJpaDao.findByEmail(memberEmail)
+            .ifPresent(user -> {
+                throw new IllegalArgumentException(ErrorMessage.EMAIL_ALREADY_EXISTS_MSG);
+            });
```
추후에 로직을 추가할 때 재사용할 수 있도록 로직들을 메서드로 따로 구성해봤습니다.

서비스 개수가 많아지고 이렇게 확인해야할 점들이 늘어가며 쿼리의 종류가 증가했는데요
보시다시피 각각의 쿼리들이 member, wishlist, product 와 같이 다양한 엔티티의 DAO를 접근하고 있습니다.
아직은 로직이 복잡하지 않아 괜찮지만, 더 커지게 된다면 이렇게 product DAO를 참조해야하는 로직이 member 서비스에 존재하듯 흩뿌려져 있는 것이 유지보수하기에도 어렵고 코드의 일관성도 좀 떨어지고 중복을 야기할수도 있을 것 같은데

예를들면 product DAO에서 해당 상품이 존재하는지 확인하는 쿼리 로직과 같은 경우는 member서비스에서 수행하지 않고 product서비스에 위임하도록 메서드들을 구성하는게 좀 더 좋은 구조일것 같은데
이 부분에 대한 의견 주시면 감사하겠습니다!

jaymarystella replied:
좋은 고민해주셨네요~
넵 저도 그 부분에 동의합니다.  service의 덩치가 큰데 일부메서드만 사용한다면 응집력이 떨어지기떄문에 repository를 바로 가져도 괜찮다고 생각합니다.

5win commented on src/main/java/gift/dto/OptionDto.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Min(value = 1, message = ErrorMessage.OPTION_QUANTITY_INVALID_MSG)
+    @Max(value = 99_999_999, message = ErrorMessage.OPTION_QUANTITY_INVALID_MSG)
+    private int quantity;
```
옵션 수량이 최대 1억개 미만이라 int로 선언하긴 했습니다만
만약 클라이언트에서 int형은 물론 long형도 벗어나도록 데이터가 들어온다면 오류가 발생할 수 있을 것 같단 생각이 듭니다.

물론 프론트단에서 유효성 검사를 한 뒤에 보내주면 되겠지만 저번에 말씀해주신 것 처럼 각각의 레이어에서 모두 유효성 검사를 하기 위해서는 왠지 String으로 받아서 검사를 해야할 것 같다고 생각이 되는데 바꾸는 것이 좋을까요?

jaymarystella replied:
일단 협의된 값이기때문에 큰 문제는 없어보입니다. 그리고 저 값은 서비스도중에 정책적인부분과 함께 충분히 바꿔질 수 있는 부분이라 괜찮습니다.

다만 매우 좋은 지적을 해주셨는데요, 보통은 다른서버들에서 넘어오는 id값을 저런식으로 정의하다가 장애가 터지는경우가 있는데요, 제한된 개수의 도메인이라고 생각되어도 외부에서 들어오는 아이디정보들은 안전하게 String으로 받고 파싱하는방식으로 구현하시는게 안전할 수 있습니다.

5win commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -17,44 +20,85 @@ public class ProductService {
+        assertProductNotDuplicate(productRequest.getName());
+        Category category = findCategoryByIdOrElseThrow(productRequest.getCategoryId());
+        Product savedProduct = productJpaDao.save(new Product(productRequest, category));
+        addOptions(productRequest, savedProduct);
```
상품 등록 시에 옵션도 함께 추가하도록 하기 위해서 로직을 이렇게 구성했습니다.
사실 그렇게 마음에 들지는 않는데 마땅히 더 좋은 방법이 떠오르진 않네요..

상품 등록과 옵션 추가가 마치 트랜잭션의 원자성?을 보장하듯 "모두 추가되거나 아예 하나도 안되거나" 를 보장하도록 하고싶은데
옵션 추가 전에 product의 save() 결과가 null인지 확인하는 것으로 충분할지 잘 모르겠습니다.

혹시 떠오르는 방법이 있다면 조언해주시면 감사하겠습니다 :)

jaymarystella replied:
transactional에 대해서 한번 찾아보면 어떨까요?

5win replied:
> transactional에 대해서 한번 찾아보면 어떨까요?

감사합니다! 덕분에 transactional에 대한 이해도가 조금 더 향상된 것 같습니다

5win commented on src/main/java/gift/service/OptionService.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    public void deleteOption(Long productId, Long optionId) {
+        findOptionByIdOrElseThrow(optionId);
+        if (findAllOptionsByProductId(productId).size() == 1) {
```
상품은 옵션을 반드시 가져야하기 때문에 1개밖에 없다면 삭제하지 못하도록 했습니다.
때문에 findAll 쿼리가 추가적으로 수행되는데, 만약 product 엔티티에 옵션 개수를 컬럼으로 가지는 것은 어떨까요?

동시성과 일치성 문제가 발생할 수도 있을 것 같은데 만약 옵션 개수 변수를 추가해서 관리하고자 한다면
동시성과 일치성 부분을 어떻게 보장할 수 있는지 키워드나 방향성? 정도 주시면 도움이 될 것 같습니다!

jaymarystella replied:
좋은 고민 해주신거같습니다.
개인적으로는 수량컬럼까지 관리하게된다면 추가나 삭제때도 같이 관리되어야한다는 단점이 있을것같네요. (전문 이커머스 서비스에서 수량관리 하는 방식은 제가 잘 모르겠습니다. 메모리에서 먼저까고 rds를 추후에 맞춰주는방식을 쓴다고했던거같기도하고.. 이부분은 서치를 좀 해보시면 좋을것같습니다) 

옵션테이블에서 항상 검사해주는방식과 수량관리 방식의 트레이드오프를 고려해주시면 될것같습니다.

jaymarystella replied:
동시성문제를 고려한다면 db 격리수준과 락에 대해서 찾아보시면 됩니다.

5win commented on src/test/java/gift/entity/CategoryTest.java:
```diff
@@ -0,0 +1,21 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class CategoryTest {
```
getter와 같은 메서드는 제외하고, 그외의 로직을 가지는 메서드에 대해선 테스트코드를 작성해보았습니다!

jaymarystella commented on src/main/java/gift/controller/OptionController.java:
```diff
@@ -0,0 +1,46 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/products/product/{productId}/options")
```
product는 빼도 되지 않을까요?

5win replied:
> product는 빼도 되지 않을까요?

넵 아직 반영하진 못했는데 추후에 반영하도록 하겠습니다!

jaymarystella commented on src/main/java/gift/controller/OptionController.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @PutMapping
+    public ResponseEntity<String> editOption(@RequestBody @Valid OptionDto optionDto) {
```
productId가 쓰이는곳도 있고 쓰이지 않는곳도 있네요. 개선해볼 수 있을까요?

5win replied:
> productId가 쓰이는곳도 있고 쓰이지 않는곳도 있네요. 개선해볼 수 있을까요?

안그래도 옵션 등록과 수정에 필요한 데이터가 좀 달라 고민하긴 했는데
이번 기회에 OptionSaveRequest와 OptionEditRequest로 나누어서 로직에 맞게 적용시켜주었습니다.

다만, 슬슬 동일 도메인에 대한 여러 DTO도 생겨나고 도메인의 개수가 많아지다보니
패키지 구조를 도메인 중심으로 묶는 방향에 대해서도 고려해볼 수 있을 것 같은 생각이 듭니다.

jaymarystella commented on src/main/java/gift/dto/OptionDto.java:
```diff
@@ -0,0 +1,58 @@
+
+public class OptionDto {
+
+    private Long id;
```
이부분도 필요한가요?

5win replied:
> 이부분도 필요한가요?

DTO를 분리하면서 id가 필요없는 부분은 제거해봤습니다!

jaymarystella commented on src/main/java/gift/exceptionHandler/GlobalExceptionHandler.java:
```diff
@@ -31,6 +33,7 @@ public ResponseEntity<String> inputValidException(MethodArgumentNotValidExceptio
      */
     @ExceptionHandler(NoSuchElementException.class)
     public ResponseEntity<String> noSuchElementExceptionHandler(NoSuchElementException e) {
+        System.out.println(e.getStackTrace()[0]);
```
로깅을 사용해보면 어떨까요?

5win replied:
> 로깅을 사용해보면 어떨까요?

반영하여 SLF4J 를 사용해 로깅해보았습니다!

jaymarystella commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -17,44 +20,85 @@ public class ProductService {
+                    throw new IllegalArgumentException(ErrorMessage.OPTION_NAME_ALREADY_EXISTS_MSG);
+                });
+            optionJpaDao.save(new Option(optionDto, savedProduct));
+        });
```
bulk조회로 개선이 가능한 부분일까요?

5win replied:
> bulk조회로 개선이 가능한 부분일까요?

의도하신 바가 매 반복마다 일일이 findByNameAndProduct_Id()로 조회하지 않고, 한 번에 findAllByProduct_Id()으로 옵션 리스트를 받아서 존재하는지 확인하는 방법일 것 같다고 판단했는데 혹시 맞을까요?

일단 optionJpaDao로 일일이 조회하지 않고, Product 객체를 통해 Product 도메인에서 해당 옵션이 존재하는지 확인하도록 변경했습니다.
이 과정에서 Product객체의 List<Option>에 옵션을 추가해줌으로써 메모리 상에서 중복 여부를 판단할 수 있도록 수정했습니다.

jaymarystella replied:
넵 맞습니다. 어차피 다 가져올거라면 단건보다는 다건형태가 더 낫습니다. 다만 쿼리시 필터할 수 있는 정보들은 최대한필터하는게좋습니다.

jaymarystella commented on src/main/java/gift/service/OptionService.java:
```diff
@@ -0,0 +1,104 @@
+    public void addOption(OptionDto optionDto) {
+        Product product = findProductByIdOrElseThrow(optionDto.getProductId());
+        assertOptionNotDuplicate(optionDto.getName(), optionDto.getProductId());
+        optionJpaDao.save(new Option(optionDto, product));
```
상품에서 옵션을 조회할 수 있는 연관관계를 맺어주셨는데 optionRepository에서 조회하고있는거같은데요, 정의해주신 도메인 클래스내에서 해결할수 있는 방법으로 고민해보시면 좋을거같아요. 다른부분들도 한번씩 확인해주세요.

jaymarystella replied:
네이밍에 orElseThrow는 없어도 될거같은데 어떻게 생각하시나요?

5win replied:
> 네이밍에 orElseThrow는 없어도 될거같은데 어떻게 생각하시나요?

예외를 반환한다는 의미로 붙이긴 했는데 너무 긴 느낌이 들긴 하네요..
저도 작성하다보니 너무 긴 느낌이 들어 일단 줄이긴 했습니다.

jaymarystella reviewed (approved):
(잘못눌렀네요)

jaymarystella reviewed (changes requested):
안녕하세요. 리뷰반영 잘해주셨습니다.
고민도 많이 해주셨네요. 제가 질문드린 부분에 답 남겨두었는데 부족한부분있으면 다시 말씀주세요!

5win committed:
feat: Add transactional annotation to assure atomicity at addProduct method
- 상품 등록과 옵션 추가 트랜잭션의 원자성을 보장하도록 Transactional 적용.

5win committed:
refact: Split OptionDto to OptionSaveRequest, OptionEditReqeust, OptionResponse

5win committed:
refact: Delegate option duplicate verification logic to the product entity
- Product entity
	- addOption() 추가.
	- isOptionDuplicate() 추가.
- OptionService의 assertOptionNotDuplicate() 제거.

5win committed:
refact: Delegate getting option and option size to the product entity

5win committed:
test: Add test code for add product with multiple options

5win committed:
refact: Change to use lsf4j for logging

5win committed:
refact: Move domain logic from service layer to domain layer
- 전반적으로 Option 접근 로직을 Product를 통해 접근하도록 수정.

5win committed:
chore: Remove OrElseThrow from method name

5win committed:
test: Add test code for Product entity
- updateOption()과 canDeleteOption()은 아직 테스트하지 못함.

5win commented on src/main/java/gift/controller/OptionController.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<String> addOption(@RequestBody @Valid OptionSaveRequest saveRequest) {
```
저장과 수정에 따라 다르게 값을 받기 때문에 OptionSaveRequest와 OptionEditRequest 그리고 OptionResponse로 분리해보았습니다.

5win commented on src/main/java/gift/dto/OptionEditRequest.java:
```diff
@@ -0,0 +1,53 @@
+
+    private Long id;
+
+    @NotBlank(message = ErrorMessage.OPTION_NAME_NOT_BLANK_MSG)
```
OptionEditRequest와 OptionSaveRequest에서 공통된 변수에 대해서 동일한 검증 로직이 들어가서 중복이 발생하는데, 아직 개선하진 못했습니다.
방법을 생각해보고 추후 보완해보도록 하겠습니다!

5win commented on src/main/java/gift/entity/Product.java:
```diff
@@ -70,17 +77,66 @@ public Category getCategory() {
+    /**
+     * 옵션은 상품에 종속된 하위 개념이므로 Product에서 Option을 수정하도록 명령.
+     */
+    public void updateOption(OptionEditRequest optionEditRequest) {
```
슬랙으로 질문드린 사항에 대한 답변을 보고
Product로 도메인 로직을 최대한 옮겨보았습니다.

말씀주신 것처럼 옵션은 상품의 하위 개념으로 볼 수 있기 때문에 서비스가 아닌 상품 도메인에게 옵션을 수정하는 작업을 옵션 도메인에게 수행시키도록 했습니다.

5win commented on src/main/java/gift/entity/Product.java:
```diff
@@ -70,17 +77,66 @@ public Category getCategory() {
+            );
+    }
+
+    public boolean canDeleteOption(Long optionId) {
```
옵션 삭제 로직도 도메인을 통해 수행하도록 하려 했으나, 연관관계의 주인이 옵션인 관계로 실패했습니다.

대신 삭제해도 되는지를 상품에게 묻고 서비스에서 OptionJpaDao 접근을 통해 삭제하도록 했습니다.

5win commented on src/main/java/gift/exceptionHandler/GlobalExceptionHandler.java:
```diff
@@ -11,13 +14,16 @@
 @ControllerAdvice
 public class GlobalExceptionHandler {
 
+    private final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);
```
피드백 반영하여 lsf4j를 사용해서 로깅했습니다.

5win committed:
refact: Remove not used method in OptionJpaDao

5win commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -1,60 +1,108 @@
+                throw new IllegalArgumentException(ErrorMessage.OPTION_NAME_DUPLICATE_MSG);
+            }
+
+            savedProduct.addOption(option);
```
Product 객체의 List<Option>에 객체를 저장해주므로써 조회 쿼리 없이 중복을 확인할 수 있도록 변경했습니다.

5win commented on src/test/java/gift/entity/ProductTest.java:
```diff
@@ -0,0 +1,55 @@
+
+        assertThat(product.isOptionNameDuplicate(option2)).isTrue();
+
+    }
```
몇 가지 추가한 메서드를 테스트하지 못했습니다

Option이 id를 입력받지 않아 로직 상 DB에 저장 없이는 getId()를 호출할 수 없어 발생한 부분도 있고,
private 메서드는 리플렉션을 통해 테스트하려 했으나 이 또한 리플렉션과 mockito를 함께 사용할 수 없는 것 같더라구요.
powerMockito라는게 있었는데 추가적인 의존성이 필요해서 일단 진행하진 않았습니다.

id의 경우 setId()를 선언해서 테스트할 수 있을 것 같은데, 테스트만을 위해서 setId()를 추가하는게 맞는지 잘 모르겠습니다.
그 외 방법은 레포지토리를 주입받아 하는 방법이 있긴 하겠지만 엔티티를 테스트하는데 너무 돌아가는 것 같아 고민을 좀 더 해보겠습니다.

jaymarystella replied:
private메서드에 대한 테스트는 어떻게 해야할지 조금더 고민해보셔요 !

jaymarystella reviewed (approved):
안녕하세요 리뷰 반영하느라 고생많으셨습니다.
고민도 많이 해주셨네요! 다음단계에서 뵙겠습니다.
